aptos/wormhole/sources/external_address.move

```diff
@@ -0,0 +1,139 @@
+/// 32 byte, left-padded address representing an arbitrary address, to be used in VAAs to
+/// refer to addresses.
+module wormhole::external_address {
+    use aptos_framework::vector;
+
+    const E_VECTOR_TOO_LONG: u64 = 0;
+
+    struct ExternalAddress has drop, copy, store {
+        external_address: vector<u8>,
+    }
+
+    public fun get_bytes(e: &ExternalAddress): vector<u8> {
+        e.external_address
+    }
+
+    public entry fun pad_left_32(input: &vector<u8>): vector<u8>{
+        let len = vector::length<u8>(input);
+        assert!(len <= 32, E_VECTOR_TOO_LONG);
+        let ret = vector::empty<u8>();
+        let zeros_remaining = 32 - len;
+        while (zeros_remaining > 0){
+            vector::push_back<u8>(&mut ret, 0);
+            zeros_remaining = zeros_remaining - 1;
+        };
+        vector::append<u8>(&mut ret, *input);
+        ret
+    }
+
+    public fun left_pad(s: &vector<u8>): ExternalAddress {
+        let padded_vector = pad_left_32(s);
+        ExternalAddress { external_address: padded_vector}
+    }
+
+    fun take_32(bytes: vector<u8>): vector<u8> {
+        let res = vector::empty<u8>();
+        let length = vector::length<u8>(&bytes);
+        let i = 0;
+        while (i < length) {
+            if ((length-i)<=32){
+                vector::push_back(&mut res, *vector::borrow(&bytes, i));
+            };
+            i = i +1;
+        };
+        res
+    }
+
+    public fun from_vector(bytes: vector<u8>): ExternalAddress {
+        left_pad(&take_32(bytes))
```

Comment:
I'm tempted to suggest that we should not allow longer than 32 bytes coming in here, and only do left-padding.

In contrast, for `String32` it's reasonable to expect truncation semantics because `String` represents free-form text, which we want to squeeze into our payloads, but an `ExternalAddress` should never come in longer than 32 bytes, because then truncation loses some information which becomes unrecoverable. At best, this is a denial of service, at worst it's a security risk.

Instead, if the 'address' is longer, then we need a preprocessing mechanism for fitting it into 32 bytes first, like the hashing we do for native token addresses. But that should not be the responsibility of the `ExternalAddress` module I think (as the preprocessing step is domain-specific). So it should just reject anything longer than 32 bytes.

Reply:
got rid of take_32